### ISSUE TITLE
Upgraded AngularJS component broken transclusion of Angular.io component under ng-if

### ISSUE THREAD

**Issue by zuzusik** [angular/angular#29026]
# 🐞 bug report

### Affected Package

 The issue is caused by package @angular/upgrade/static

### Is this a regression?

 Not sure

### Description

 

With situation described below when `ng-if` adds content, transclusion happens correctly only once - next times ng-if adds content, components on the transcluded content are not initialized

```html
<aio-app> <!-- downgraded aio component -->
  <ajs-show-hide> <!--  upgraded ajs component -->
    <button ng-init="shown = false" ng-click="shown = !shown">Show/Hide</button>
    <div ng-if="shown">
      <ng-transclude>
        <aio-counter> <!-- native aio component, transcluded by aio-app into ajs-show-hide -->
          <ajs-counter> <!--  upgraded ajs component -->
            <button ng-init="count = 0" ng-click="count = count + 1">Bump!</button>
            Count: {{count}}
          </ajs-counter>
        </aio-counter>
      </ng-transclude>
    </div>
  </ajs-show-hide>
</aio-app>
```

## 🔬 Minimal Reproduction

 
1. clone https://github.com/zuzusik/angular-upgrade-bug
2. run `npm install`
3. run `ng serve`
4. open http://localhost:4201/
5. click `Show/Hide` button
6. click `Bump!` button
7. ensure Count is updated
8. click `Show/Hide` button again, so that counter is hidden
9. click `Show/Hide` button again, so that counter is shown again
10. observe counter being in previous state (count shows the number of the last count visible) - this is not expected, it should be set to 0
11.  click `Bump!` button
12. counter is not updated - the component is broken

## 🔥 Exception or Error
<pre><code>

No errors
</code></pre>

## 🌍  Your Environment

**Angular Version:**
<pre><code>

Angular CLI: 7.3.3
Node: 8.12.0
OS: darwin x64
Angular: 7.2.7
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router, upgrade

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.13.3
@angular-devkit/build-angular     0.13.3
@angular-devkit/build-optimizer   0.13.3
@angular-devkit/build-webpack     0.13.3
@angular-devkit/core              7.3.3
@angular-devkit/schematics        7.3.3
@angular/cli                      7.3.3
@ngtools/webpack                  7.3.3
@schematics/angular               7.3.3
@schematics/update                0.13.3
rxjs                              6.3.3
typescript                        3.2.4
webpack                           4.29.0
</code></pre>

**Anything else relevant?**

**Comment by gkalpak** [comment]
Thx for reporting this, @zuzusik. Could you, please, create a minimal reproduction on StackBlitz? (You can use [this][1] as a starting point.)

[1]: https://stackblitz.com/edit/ngupgradestatic-playground

**Comment by zuzusik** [comment]
@gkalpak sure, here we go: https://stackblitz.com/edit/angular-puip4w?file=src%2Findex.html

ps: the template you have provided looks way too outdated - I just started from typical Angular template

**Comment by gkalpak** [comment]
I looked into this a bit (but didn't really get to the bottom of it yet). Here are some thoughts/findings:

This seems to be hitting the limits of `ngUpgrade` (having 4 levels of interleaved AngularJS/Angular components + transclusion/content projection + structural directives) :grin:

Also, note, that Angular's content projection works differently than AngularJS' transclusion (the former instantiates the component eagerly, while the latter does it lazily).

In addition to that, our [transclusion emulation][1] apparently doesn't do a great job at handling such complex scenarios.
Finally, there might be some issue with propagating `$destroy` events from AngularJS structural directives to downgraded Angular components. (I need to investigate more.)

##
There is definitely more investigation to be done. I will try to get to it in the next few days.

BTW, for reference, here is a sample implementation of the failing example (also showing when each implementation calls the several lifecycle hooks) in:
- [Pure AngularJS][2]
- [Pure Angular][3]
- [Hybrid AngularJS/Angular][4]

[1]: https://github.com/angular/angular/blob/dcafddefb8004351561fbceed02f3bf3df1bb6c9/packages/upgrade/src/common/upgrade_helper.ts#L139-L220
[2]: https://stackblitz.com/edit/ng-issue-29026-angularjs?file=app.js
[3]: https://stackblitz.com/edit/ng-issue-29026-angular?file=src%2Fmain.ts
[4]: https://stackblitz.com/edit/ng-issue-29026-angular?file=src%2Fmain.ts

**Comment by zuzusik** [comment]
> This seems to be hitting the limits of `ngUpgrade` (having 4 levels of interleaved AngularJS/Angular components + transclusion/content projection + structural directives) 😁

Are these limits documented somewhere? 

Also, it would be nice to get console errors when they are reached - otherwise it is totally not clear what is happening.

**Comment by zuzusik** [comment]
>     * [Pure Angular](https://stackblitz.com/edit/ng-issue-29026-angular?file=src%2Fmain.ts)
> 
>     * [Hybrid AngularJS/Angular](https://stackblitz.com/edit/ng-issue-29026-angular?file=src%2Fmain.ts)

both these links point to `Hybrid AngularJS/Angular` example

**Comment by zuzusik** [comment]
>     * [Hybrid AngularJS/Angular](https://stackblitz.com/edit/ng-issue-29026-angular?file=src%2Fmain.ts)

Also, this doesn't fully reproduces issue - after show/hide the counter is not reset (a part of the originally described issue), but it still works (in original issue it gets broken after first show/hide)

**Comment by gkalpak** [comment]
>> This seems to be hitting the limits of ngUpgrade (having 4 levels of interleaved AngularJS/Angular components + transclusion/content projection + structural directives) 😁
>
> Are these limits documented somewhere?
> Also, it would be nice to get console errors when they are reached - otherwise it is totally not clear what is happening.

No, we don't know what the limits are (nor are they intentional) :grin: 

##
> both these links point to Hybrid AngularJS/Angular example

Ooops, fixed.

##
>> [Hybrid AngularJS/Angular](https://stackblitz.com/edit/ng-issue-29026-angular?file=src%2Fmain.ts)
>
> Also, this doesn't fully reproduces issue - after show/hide the counter is not reset (a part of the originally described issue), but it still works (in original issue it gets broken after first show/hide)

No, this is pure Angular implementation and has no problem. The problem is with ngUpgrade (and the specific combination of nesting components/transcluding/structural directives).

**Comment by zuzusik** [comment]
Just in case: encountered this recently on Angular 11.0.2 - the bug is still there.

**Comment by JeanMeche** [comment]
Without further reports in the last 4 year, I'll considert this case closed.
